File: src/main/java/br/com/matheuspadilha/tasks/event/TaskNotificationProducer.java
package br.com.matheuspadilha.tasks.event;

import br.com.matheuspadilha.tasks.model.Task;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Mono;

@Component
public class TaskNotificationProducer {

    private static final Logger LOGGER = LoggerFactory.getLogger(TaskNotificationProducer.class);

    @Value("${kafka.task.notification.output}")
    private String topic;
    private final KafkaTemplate<Object, Object> template;

    public TaskNotificationProducer(KafkaTemplate<Object, Object> template) {
        this.template = template;
    }

    public Mono<Task> sendNotification(Task task) {
        return Mono.just(task)
                .map(it -> this.template.send(topic, task))
                .map(it -> task)
                .doOnNext(it -> LOGGER.info("Task notification send successfully"));
    }
}

File: src/main/kotlin/br/com/matheuspadilha/tasks/event/TaskNotificationProducer.kt
package br.com.matheuspadilha.tasks.event

import br.com.matheuspadilha.tasks.model.Task
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.stereotype.Component
import reactor.core.publisher.Mono

@Component
class TaskNotificationProducer(
    private val template: KafkaTemplate<Any, Any>
) {

    fun send(task: Task): Mono<Task> =
        Mono.just(task)
            .map { template.send(TASK_NOTIFICATION_V1_PRODUCER, it) }
            .map { task }
            .doOnNext { LOGGER.info("Task notification send successfully") }

    companion object {
        private val LOGGER: Logger = LoggerFactory.getLogger(this::class.java)
        private const val TASK_NOTIFICATION_V1_PRODUCER = "taskEventProducer-out-0"
    }
}